### PR TITLE
Issue 5308: Make max read length configurable in PravegaRequestProcessor

### LINKED ISSUE #5308: Change various performance parameters from constants to use run-time properties
**Problem description**
Some parameters are hard-coded as compile-time constants, making it difficult to test different values.

**Problem location**
For example:
https://github.com/pravega/pravega/blob/7e8a26133281a86ca8823c4665fd7d2fb836eeb1/segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java#L134

**Suggestions for an improvement**
Get parameter values from Java properties.

### PR DESCRIPTION
**Pull request by RaulGracia**
**Change log description**  
Added a parameter (`pravegaservice.request.read.length.max`) to configure the max size of reads that the `PravegaRequestProcessor` issues to the cache. Also, this PR aligns the default values of `pravegaservice.request.read.length.max` and `readindex.storageRead.alignment` according to the max read length that clients can issue (i.e., 10MB).

**Purpose of the change**  
Fixes #5308.

**What the code does**  
To improve the performance of historical reads when just few segments are available in parallel, we need to issue larger reads to LTS. However, one hardcoded variable (`MAX_READ_SIZE`, set to 2MB) in `PravegaRequestProcessor` was preventing large reads from clients (up to 10MB) to take effect. This PR basically replaces such a hardcoded variable by a configurable parameter.

In addition, to better exploit large client reads issues to LTS, we aligned the values of `pravegaservice.request.read.length.max` (reads from `PravegaRequestProcessor` to the cache) and `readindex.storageRead.alignment` (prefetch from LTS to cache) according to the max read size of clients (10MB at the moment, as defined in `SegmentInputStreamImpl`).

**How to verify it**  
All tests should be working as before with the new configuration parameter.

### CODE REVIEW

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/ServiceStarter.java [hunk 1]
```diff
@@ -118,7 +118,9 @@ public void start() throws Exception {
                                                       this.serviceConfig.getListeningPort(), service, tableStoreService,
                                                       autoScaleMonitor.getStatsRecorder(), autoScaleMonitor.getTableSegmentStatsRecorder(),
                                                       tokenVerifier, this.serviceConfig.getCertFile(), this.serviceConfig.getKeyFile(),
-                                                      this.serviceConfig.isReplyWithStackTraceOnError(), serviceBuilder.getLowPriorityExecutor());
+                                                      this.serviceConfig.isReplyWithStackTraceOnError(),
```
Comment (andreipaduroiu) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/ServiceStarter.java:121]: Any chance we could do something to this constructor? It's getting out of hands. Could we add a `@Builder` tag to it?

If we do add a `@Builder`, then we could potentially get rid of the other 1-2 constructors next to it and supply defaults via the builder object.

Not really related to this PR, but every time I had to modify this area of the code I found myself having to modify a good number of other files too.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaConnectionListener.java [hunk 2]
```diff
@@ -110,7 +110,7 @@
     @VisibleForTesting
     public PravegaConnectionListener(boolean enableTls, int port, StreamSegmentStore streamSegmentStore, TableStore tableStore, ScheduledExecutorService tokenExpiryExecutor) {
         this(enableTls, false, "localhost", port, streamSegmentStore, tableStore, SegmentStatsRecorder.noOp(), TableSegmentStatsRecorder.noOp(),
-                new PassingTokenVerifier(), null, null, true, tokenExpiryExecutor);
+                new PassingTokenVerifier(), null, null, true, 10 * 1024 * 1024, tokenExpiryExecutor);
```
Comment (andreipaduroiu) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaConnectionListener.java:113]: Such as this constructor. That value could be a nice default in a builder, along with all the nulls and no-op classes.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaConnectionListener.java [hunk 3]
```diff
@@ -156,6 +157,7 @@ public PravegaConnectionListener(boolean enableTls, boolean enableTlsReload, Str
         }
         this.replyWithStackTraceOnError = replyWithStackTraceOnError;
         this.connectionTracker = new ConnectionTracker();
+        this.maxReadLength = maxReadLength;
```
Comment (andreipaduroiu) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaConnectionListener.java:160]: Please add a `Preconditions.checkArgument` to validate that this is at least X KB. (X == 10?).

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java [hunk 4]
```diff
@@ -167,17 +167,19 @@ public PravegaRequestProcessor(StreamSegmentStore segmentStore, TableStore table
      * @param tableStatsRecorder A TableSegmentStatsRecorder for Metrics for Table Segments.
      * @param tokenVerifier  Verifier class that verifies delegation token.
      * @param replyWithStackTraceOnError Whether client replies upon failed requests contain server-side stack traces or not.
+     * @param maxReadLength  Max read length that the {@link PravegaRequestProcessor} can issue to the cache.
      */
     PravegaRequestProcessor(StreamSegmentStore segmentStore, TableStore tableStore, ServerConnection connection,
                             SegmentStatsRecorder statsRecorder, TableSegmentStatsRecorder tableStatsRecorder,
-                            DelegationTokenVerifier tokenVerifier, boolean replyWithStackTraceOnError) {
+                            DelegationTokenVerifier tokenVerifier, boolean replyWithStackTraceOnError, int maxReadLength) {
```
Comment (andreipaduroiu) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java:174]: Another candidate for `@Builder` :)

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java [hunk 5]
```diff
@@ -167,17 +167,19 @@ public PravegaRequestProcessor(StreamSegmentStore segmentStore, TableStore table
...
         this.segmentStore = Preconditions.checkNotNull(segmentStore, "segmentStore");
         this.tableStore = Preconditions.checkNotNull(tableStore, "tableStore");
         this.connection = Preconditions.checkNotNull(connection, "connection");
         this.tokenVerifier = Preconditions.checkNotNull(tokenVerifier, "tokenVerifier");
         this.statsRecorder = Preconditions.checkNotNull(statsRecorder, "statsRecorder");
         this.tableStatsRecorder = Preconditions.checkNotNull(tableStatsRecorder, "tableStatsRecorder");
         this.replyWithStackTraceOnError = replyWithStackTraceOnError;
+        this.maxReadLength = maxReadLength;
```
Comment (andreipaduroiu) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java:182]: Validate.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java [hunk 6]
```diff
@@ -194,7 +196,7 @@ public void readSegment(ReadSegment readSegment) {
             return;
         }
 
-        final int readSize = min(MAX_READ_SIZE, max(TYPE_PLUS_LENGTH_SIZE, readSegment.getSuggestedLength()));
+        final int readSize = min(maxReadLength, max(TYPE_PLUS_LENGTH_SIZE, readSegment.getSuggestedLength()));
```
Comment (andreipaduroiu) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessor.java:199]: `MathHelpers.minMax`

File: segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessorAuthFailedTest.java [hunk 7]
```diff
@@ -36,7 +36,7 @@ public void setUp() throws Exception {
                 TableSegmentStatsRecorder.noOp(),
                 (resource, token, expectedLevel) -> {
                     throw new InvalidTokenException("Token verification failed.");
-                }, false);
+                }, false, 10 * 1024 * 1024);
```
Comment (andreipaduroiu) [inline segmentstore/server/host/src/test/java/io/pravega/segmentstore/server/host/handler/PravegaRequestProcessorAuthFailedTest.java:39]: why not use the constant here?

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/ServiceConfig.java [hunk 8]
```diff
@@ -355,6 +363,7 @@ private ServiceConfig(TypedProperties properties) throws ConfigurationException
         this.cachePolicy = new CachePolicy(cachePolicyMaxSize, cachePolicyTargetUtilization, cachePolicyMaxUtilization,
                 Duration.ofSeconds(cachePolicyMaxTime), Duration.ofSeconds(cachePolicyGenerationTime));
         this.replyWithStackTraceOnError = properties.getBoolean(REPLY_WITH_STACK_TRACE_ON_ERROR);
+        this.requestProcessorMaxReadLength = properties.getInt(REQUEST_PROCESSOR_MAX_READ_LENGTH);
```
Comment (andreipaduroiu) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/store/ServiceConfig.java:366]: Validate that it is positive. And optionally greater than some min value. Let's make it consistent with PravegaRequestProcessor/Listener

**Review by andreipaduroiu — CHANGES_REQUESTED** [review]

**Comment by andreipaduroiu** [comment]
I have certain reservations when it comes to increasing these values ten-fold. I like it that we made them configurable, but I believe that we should keep the defaults as they are. We can set them to higher values at deployment time, and if this proves not to introduce instability in the system, we can choose to update the defaults as well.

**Comment by RaulGracia** [comment]
Closing this one. If needed, I will raise it again.

**Pull request closed without merging**